<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Feature plots | SCpubr - Complete reference</title>
  <meta name="description" content="General wrapper of known functions of Seurat, alongside other common visualizations in Single Cell experiments." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Feature plots | SCpubr - Complete reference" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="General wrapper of known functions of Seurat, alongside other common visualizations in Single Cell experiments." />
  <meta name="github-repo" content="enblacar/SCpubr" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Feature plots | SCpubr - Complete reference" />
  
  <meta name="twitter:description" content="General wrapper of known functions of Seurat, alongside other common visualizations in Single Cell experiments." />
  

<meta name="author" content="Enrique Blanco Carmona" />


<meta name="date" content="2022-02-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dim-plots.html"/>
<link rel="next" href="nebulosa-plots.html"/>
<script src="assets/header-attrs-2.11/header-attrs.js"></script>
<script src="assets/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="assets/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="assets/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./"><b>SCpubr - Complete reference guide</b></a></li>
<li><a href="https://github.com/enblacar/SCpubr" target="blank"><b>Link to Github repository</b></a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Aim of the package</a></li>
<li class="chapter" data-level="2" data-path="installation-guide.html"><a href="installation-guide.html"><i class="fa fa-check"></i><b>2</b> Installation guide</a></li>
<li class="chapter" data-level="3" data-path="dataset.html"><a href="dataset.html"><i class="fa fa-check"></i><b>3</b> Dataset</a></li>
<li class="chapter" data-level="4" data-path="dim-plots.html"><a href="dim-plots.html"><i class="fa fa-check"></i><b>4</b> Dim plots</a>
<ul>
<li class="chapter" data-level="4.1" data-path="dim-plots.html"><a href="dim-plots.html#regular-dimplots"><i class="fa fa-check"></i><b>4.1</b> Regular DimPlots</a></li>
<li class="chapter" data-level="4.2" data-path="dim-plots.html"><a href="dim-plots.html#highlighting-cells"><i class="fa fa-check"></i><b>4.2</b> Highlighting cells</a></li>
<li class="chapter" data-level="4.3" data-path="dim-plots.html"><a href="dim-plots.html#splitting-by-a-category"><i class="fa fa-check"></i><b>4.3</b> Splitting by a category</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="feature-plots.html"><a href="feature-plots.html"><i class="fa fa-check"></i><b>5</b> Feature plots</a>
<ul>
<li class="chapter" data-level="5.1" data-path="feature-plots.html"><a href="feature-plots.html#multiple-features"><i class="fa fa-check"></i><b>5.1</b> Multiple features</a></li>
<li class="chapter" data-level="5.2" data-path="feature-plots.html"><a href="feature-plots.html#working-with-subsets-of-cells"><i class="fa fa-check"></i><b>5.2</b> Working with subsets of cells</a></li>
<li class="chapter" data-level="5.3" data-path="feature-plots.html"><a href="feature-plots.html#splitting-the-featureplot-by-a-variable"><i class="fa fa-check"></i><b>5.3</b> Splitting the FeaturePlot by a variable</a></li>
<li class="chapter" data-level="5.4" data-path="feature-plots.html"><a href="feature-plots.html#modify-color-maps"><i class="fa fa-check"></i><b>5.4</b> Modify color maps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="nebulosa-plots.html"><a href="nebulosa-plots.html"><i class="fa fa-check"></i><b>6</b> Nebulosa plots</a>
<ul>
<li class="chapter" data-level="6.1" data-path="nebulosa-plots.html"><a href="nebulosa-plots.html#modify-color-maps-1"><i class="fa fa-check"></i><b>6.1</b> Modify color maps</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="bee-swarm-plots.html"><a href="bee-swarm-plots.html"><i class="fa fa-check"></i><b>7</b> Bee Swarm plots</a>
<ul>
<li class="chapter" data-level="7.1" data-path="bee-swarm-plots.html"><a href="bee-swarm-plots.html#using-categorical-variables"><i class="fa fa-check"></i><b>7.1</b> Using categorical variables</a></li>
<li class="chapter" data-level="7.2" data-path="bee-swarm-plots.html"><a href="bee-swarm-plots.html#using-continuous-variables"><i class="fa fa-check"></i><b>7.2</b> Using continuous variables</a></li>
<li class="chapter" data-level="7.3" data-path="bee-swarm-plots.html"><a href="bee-swarm-plots.html#modify-color-maps-for-continuous-variables"><i class="fa fa-check"></i><b>7.3</b> Modify color maps for continuous variables</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="violin-plots.html"><a href="violin-plots.html"><i class="fa fa-check"></i><b>8</b> Violin plots</a>
<ul>
<li class="chapter" data-level="8.1" data-path="violin-plots.html"><a href="violin-plots.html#violin-plots-as-a-means-of-qc"><i class="fa fa-check"></i><b>8.1</b> Violin plots as a means of QC</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="dot-plots.html"><a href="dot-plots.html"><i class="fa fa-check"></i><b>9</b> Dot plots</a>
<ul>
<li class="chapter" data-level="9.1" data-path="dot-plots.html"><a href="dot-plots.html#querying-multiple-features-at-once."><i class="fa fa-check"></i><b>9.1</b> Querying multiple features at once.</a></li>
<li class="chapter" data-level="9.2" data-path="dot-plots.html"><a href="dot-plots.html#clustering-the-identities."><i class="fa fa-check"></i><b>9.2</b> Clustering the identities.</a></li>
<li class="chapter" data-level="9.3" data-path="dot-plots.html"><a href="dot-plots.html#inverting-the-axes."><i class="fa fa-check"></i><b>9.3</b> Inverting the axes.</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bar-plots.html"><a href="bar-plots.html"><i class="fa fa-check"></i><b>10</b> Bar plots</a>
<ul>
<li class="chapter" data-level="10.1" data-path="bar-plots.html"><a href="bar-plots.html#introducing-a-second-variable-to-group-the-feature-by"><i class="fa fa-check"></i><b>10.1</b> Introducing a second variable to group the feature by</a></li>
<li class="chapter" data-level="10.2" data-path="bar-plots.html"><a href="bar-plots.html#reordering-columns-based-on-a-single-entity"><i class="fa fa-check"></i><b>10.2</b> Reordering columns based on a single entity</a></li>
<li class="chapter" data-level="10.3" data-path="bar-plots.html"><a href="bar-plots.html#adding-custom-color-scales."><i class="fa fa-check"></i><b>10.3</b> Adding custom color scales.</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SCpubr - Complete reference</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-plots" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Feature plots</h1>
<p>Another broadly used function in <code>Seurat</code> is <code>Seurat::FeaturePlot()</code>. It is basically the counterpart of <code>Seurat::DimPlot()</code> which instead of coloring the cells based on a categorical color scale, it uses a continuous one, according to a variable provided by the user. This can range from gene expression, to metadata variables such as the number of genes, or even values such as PC_1. This is how it looks, plotting the number of UMIs:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="feature-plots.html#cb26-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;nCount_RNA&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-29"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-29-1.png" alt="Seurat FeaturePlot" width="576" />
<p class="caption">
Figure 5.1: Seurat FeaturePlot
</p>
</div>
<p>First thing that stands out, it that the color scale could be changed to one with two very opposite ends, so that identifying regions with low and high UMIs become an easier task. Apart from that, we want apply the same treatment as in <code>do_DimPlot()</code> section.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="feature-plots.html#cb27-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;nCount_RNA&quot;</span>) <span class="sc">+</span> </span>
<span id="cb27-2"><a href="feature-plots.html#cb27-2" aria-hidden="true" tabindex="-1"></a>  Seurat<span class="sc">::</span><span class="fu">NoAxes</span>() <span class="sc">+</span> </span>
<span id="cb27-3"><a href="feature-plots.html#cb27-3" aria-hidden="true" tabindex="-1"></a>  viridis<span class="sc">::</span><span class="fu">scale_color_viridis</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-30"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-30-1.png" alt="Seurat FeaturePlot, with new color scale and general modifications" width="576" />
<p class="caption">
Figure 5.2: Seurat FeaturePlot, with new color scale and general modifications
</p>
</div>
<p>This way, we can observe that most of the clusters have a really similar number of UMIs, despite selected ones which have up to 4 times more. These could very well be doublets. This is <code>SCpubr::do_FeaturePlot()</code> variation:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="feature-plots.html#cb28-1" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Number of UMIs&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-31"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-31-1.png" alt="SCpubr FeaturePlot, with minor modifications" width="576" />
<p class="caption">
Figure 5.3: SCpubr FeaturePlot, with minor modifications
</p>
</div>
<p>By default, <code>SCpubr</code> enlarges the dots a bit and makes the legend bold, to better stand out.</p>
<div id="multiple-features" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Multiple features</h2>
<p><code>SCpubr::do_FeaturePlot()</code> can tackle multiple queries at the same time. Plots can be rearranged as desired with <code>ncol</code> parameter:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="feature-plots.html#cb29-1" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;nFeature_RNA&quot;</span>, <span class="st">&quot;percent.mt&quot;</span>, <span class="st">&quot;CD14&quot;</span>), <span class="at">plot.title =</span> <span class="st">&quot;My very important features&quot;</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-32"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-32-1.png" alt="SCpubr FeaturePlot, with 4 features and 2 columns" width="576" />
<p class="caption">
Figure 5.4: SCpubr FeaturePlot, with 4 features and 2 columns
</p>
</div>
<p>With multiple queries, using <code>plot.title</code> parameter actually adds an overall title. If one wants to modify the title of each plot independently, a more complex approach is needed, which requires more R expertise, using <code>patchwork</code> package.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="feature-plots.html#cb30-1" aria-hidden="true" tabindex="-1"></a>((SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;nCount_RNA&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Plot A&quot;</span>) <span class="sc">|</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;nFeature_RNA&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Plot B&quot;</span>)) <span class="sc">/</span></span>
<span id="cb30-2"><a href="feature-plots.html#cb30-2" aria-hidden="true" tabindex="-1"></a>(SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;percent.mt&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Plot C&quot;</span>) <span class="sc">|</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Plot D&quot;</span>))) <span class="sc">+</span> </span>
<span id="cb30-3"><a href="feature-plots.html#cb30-3" aria-hidden="true" tabindex="-1"></a>patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">title =</span> <span class="st">&quot;My very important features&quot;</span>, <span class="at">theme =</span> ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">plot.title =</span> ggplot2<span class="sc">::</span><span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>)))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-33"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-33-1.png" alt="SCpubr FeaturePlot, with individual modified titles" width="576" />
<p class="caption">
Figure 5.5: SCpubr FeaturePlot, with individual modified titles
</p>
</div>
</div>
<div id="working-with-subsets-of-cells" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Working with subsets of cells</h2>
<p>One of the things that can be misleading, is that it could be the case that a very specific subset of cells are driving the ends of the color scales. Let’s say that we have identified that clusters 2, 5, and 8 are responsible of this behavior. We would like to have plot the values without taking those cells into consideration. The very first thing that comes to mind is, indeed, removing the cells completely, resulting in the following:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="feature-plots.html#cb31-1" aria-hidden="true" tabindex="-1"></a>cells.plot <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sample[, <span class="sc">!</span>(sample<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;2&quot;</span>, <span class="st">&quot;5&quot;</span>, <span class="st">&quot;8&quot;</span>))])</span>
<span id="cb31-2"><a href="feature-plots.html#cb31-2" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample[, cells.plot], <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD14&quot;</span>), <span class="at">plot.title =</span> <span class="st">&quot;My very important features&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-34"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-34-1.png" alt="SCpubr FeaturePlot, removing cells from clusters 2, 5 and 8" width="576" />
<p class="caption">
Figure 5.6: SCpubr FeaturePlot, removing cells from clusters 2, 5 and 8
</p>
</div>
<p>However, this falls back to the same problem as with <code>Seurat::DimPlot</code> with the <code>split.by</code> parameter. We lose the overall context of where the cells are. This is taken care by in <code>SCpubr::do_FeaturePlot()</code>, when using the <code>cells.highlight</code> parameter:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="feature-plots.html#cb32-1" aria-hidden="true" tabindex="-1"></a>cells.plot <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sample[, <span class="sc">!</span>(sample<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;2&quot;</span>, <span class="st">&quot;5&quot;</span>, <span class="st">&quot;8&quot;</span>))])</span>
<span id="cb32-2"><a href="feature-plots.html#cb32-2" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">cells.highlight =</span> cells.plot, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD14&quot;</span>), <span class="at">plot.title =</span> <span class="st">&quot;My very important features&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-35"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-35-1.png" alt="SCpubr FeaturePlot, masking cells from clusters 2, 5 and 8" width="576" />
<p class="caption">
Figure 5.7: SCpubr FeaturePlot, masking cells from clusters 2, 5 and 8
</p>
</div>
<p>This way, by adding the cells back and modifying the legend, we accomplish the desired effect. Furthermore, <code>SCpubr::do_FeaturePlot()</code> also accepts a vector of the identities to plot, as long as they are part of <code>levels(seurat_object)</code>. This can be achieved by using the <code>idents.highlight</code> instead.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="feature-plots.html#cb33-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-2"><a href="feature-plots.html#cb33-2" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">idents.highlight =</span> <span class="fu">levels</span>(sample)[<span class="sc">!</span>(<span class="fu">levels</span>(sample) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;2&quot;</span>, <span class="st">&quot;5&quot;</span>, <span class="st">&quot;8&quot;</span>))], <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD14&quot;</span>), <span class="at">plot.title =</span> <span class="st">&quot;My very important features&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-36"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-36-1.png" alt="SCpubr FeaturePlot, masking cells from clusters 2, 5 and 8 using idents.highlight" width="576" />
<p class="caption">
Figure 5.8: SCpubr FeaturePlot, masking cells from clusters 2, 5 and 8 using idents.highlight
</p>
</div>
</div>
<div id="splitting-the-featureplot-by-a-variable" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Splitting the FeaturePlot by a variable</h2>
<p>In some cases, we want to visualize a given feature, let’s say, the expression of <code>LYN</code> across another variable, for instance <code>seurat_clusters</code>. This can be achieved easily in <code>Seurat</code> by using the <code>split.by</code> parameter:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="feature-plots.html#cb34-1" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;LYN&quot;</span>, <span class="at">split.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-37"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-37-1.png" alt="Seurat FeaturePlot split by clusters." width="1920" />
<p class="caption">
Figure 5.9: Seurat FeaturePlot split by clusters.
</p>
</div>
<p>Apart from the general aesthetic discrepancies mentioned before, there are two main aspects missing. First, is the overall layout of the UMAP embedding. Being able to correctly locate the clusters is really important. Secondly, knowing that, by default, the color scale is shared across all the plots (achieved by using <code>keep.scale = "feature"</code>), we would like to know up to which value this color scale reaches. In this two aspects is where <code>SCpubr::do_FeaturePlot()</code> delves into:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="feature-plots.html#cb35-1" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="st">&quot;LYN&quot;</span>, <span class="at">split.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-38"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-38-1.png" alt="SCpubr FeaturePlot split by clusters." width="1920" />
<p class="caption">
Figure 5.10: SCpubr FeaturePlot split by clusters.
</p>
</div>
<p>This way, we are able to achieve a better understandable plot. Please note that this has been implemented for <code>keep.scale = "feature"</code>. Providing other options is not advisable. Furthermore, we can query multiple features, them being stacked by the different entities in <code>split.by</code>. The number of columns is predetermined to the number of entities and can not be modified. In this example, we provide each of the three possible examples of features:</p>
<ul>
<li><code>LYN</code> is a gene located in the rownames of the count matrix (`rownames(sample)``).</li>
<li><code>nCount_RNA</code> is a metadata variable located in <code>colnames(sample@meta.data)</code>.</li>
<li><code>PC_1</code> is a column of a dimensional reduction embedding, located in <code>colnames(sample@reductions[["pca"]][[]])</code></li>
</ul>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="feature-plots.html#cb36-1" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;LYN&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;PC_1&quot;</span>), <span class="at">split.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-39"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-39-1.png" alt="SCpubr FeaturePlot split by clusters using multiple features." width="1920" />
<p class="caption">
Figure 5.11: SCpubr FeaturePlot split by clusters using multiple features.
</p>
</div>
<p>With <code>SCpubr::do_FeaturePlot()</code> we can even subset the previous plot to any identities in <code>split.by</code> we are particularly interested in. The color scale also limits itself to contain only the values in the selected identities. This can be achieved by provided a vector with the identities to <code>split.by.idents</code> parameter.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="feature-plots.html#cb37-1" aria-hidden="true" tabindex="-1"></a>SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(sample, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;LYN&quot;</span>, <span class="st">&quot;nCount_RNA&quot;</span>, <span class="st">&quot;PC_1&quot;</span>), <span class="at">split.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>, <span class="at">split.by.idents =</span> <span class="fu">c</span>(<span class="st">&quot;2&quot;</span>, <span class="st">&quot;5&quot;</span>, <span class="st">&quot;7&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-40"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-40-1.png" alt="SCpubr FeaturePlot split by clusters using multiple features and selected identities." width="960" />
<p class="caption">
Figure 5.12: SCpubr FeaturePlot split by clusters using multiple features and selected identities.
</p>
</div>
</div>
<div id="modify-color-maps" class="section level2" number="5.4">
<h2><span class="header-section-number">5.4</span> Modify color maps</h2>
<p>It is possible in <code>SCpubr::do_FeaturePlot()</code> to also change the color map of the plot to one of the eight possible ones defined in <a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html">viridis</a>. This is achieved by using <code>viridis_color_map</code> parameter and the color map name or code (capital letter). Options are:</p>
<ul>
<li>A - magma color map.</li>
<li>B - inferno color map.</li>
<li>C - plasma color map.</li>
<li>D - viridis color map.</li>
<li>E - cividis color map.</li>
<li>F - rocket color map.</li>
<li>G - mako color map.</li>
<li>H - turbo color map.</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="feature-plots.html#cb38-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;A&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Magma&quot;</span>)</span>
<span id="cb38-2"><a href="feature-plots.html#cb38-2" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;B&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Inferno&quot;</span>)</span>
<span id="cb38-3"><a href="feature-plots.html#cb38-3" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;C&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Plasma&quot;</span>)</span>
<span id="cb38-4"><a href="feature-plots.html#cb38-4" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;D&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Viridis&quot;</span>)</span>
<span id="cb38-5"><a href="feature-plots.html#cb38-5" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;E&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Cividis&quot;</span>)</span>
<span id="cb38-6"><a href="feature-plots.html#cb38-6" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;F&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Rocket&quot;</span>)</span>
<span id="cb38-7"><a href="feature-plots.html#cb38-7" aria-hidden="true" tabindex="-1"></a>p7 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;G&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Mako&quot;</span>)</span>
<span id="cb38-8"><a href="feature-plots.html#cb38-8" aria-hidden="true" tabindex="-1"></a>p8 <span class="ot">&lt;-</span> SCpubr<span class="sc">::</span><span class="fu">do_FeaturePlot</span>(<span class="at">sample =</span> sample, <span class="at">features =</span> <span class="st">&quot;CD14&quot;</span>, <span class="at">viridis_color_map =</span> <span class="st">&quot;H&quot;</span>, <span class="at">plot.title =</span> <span class="st">&quot;Turbo&quot;</span>)</span>
<span id="cb38-9"><a href="feature-plots.html#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in check_viridis_color_map(viridis_color_map = viridis_color_map, : The</span></span>
<span id="cb38-10"><a href="feature-plots.html#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; selected option is not the most adequate for a continuous color scale.</span></span>
<span id="cb38-11"><a href="feature-plots.html#cb38-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-12"><a href="feature-plots.html#cb38-12" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> patchwork<span class="sc">::</span><span class="fu">wrap_plots</span>(<span class="fu">list</span>(p1, p2, p3, p4, p5, p6, p7, p8), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-13"><a href="feature-plots.html#cb38-13" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-41"></span>
<img src="SCPubr_reference_guide.html_files/figure-html/unnamed-chunk-41-1.png" alt="SCpubr FeaturePlot modifying viridis color maps." width="960" />
<p class="caption">
Figure 5.13: SCpubr FeaturePlot modifying viridis color maps.
</p>
</div>
<p>Here, we do not recommend the use of <code>turbo</code> with Feature plots, as this continuous scale goes through many different colors and, therefore, makes it harder for interpretation.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dim-plots.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="nebulosa-plots.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="assets/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": true,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": true,
"all": ["facebook", "github", "twitter", "linkedin", "whatsapp"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["SCPubr_reference_guide.html.pdf", "SCPubr_reference_guide.html.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection",
"scroll_highlight": true
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

</body>

</html>
